### ISSUE TITLE
improve code action

### ISSUE THREAD

**Issue by glepnir** [glepnir/lspsaga.nvim#116]
@akinsho Compared with nvim-lightbulb, where do you think the codeaction of lspsaga needs to be improved?

**Comment by tami5** [comment]
Oh yah 100%, I was a bit surprised that lspsaga didn't have so I readded lightblub

**Comment by akinsho** [comment]
@glepnir currently `nvim-lightbulb` provides different functionality than `lspsaga` not sure if you've tried that plugin yourself. It's probably easiest to understand if you try it but, what `lightbulb` does is when you are moving around a file if you stop on a symbol that offers a code action it will show a lightbulb similar to how you get a lightbulb in vscode to let you know that there is code action available.

That's all it does, it doesn't trigger code actions, it doesn't open up a code action floating window or anything like that. The whole point of the plugin as far as I understand it to show the user when/where they can use a code action, it doesn't get involved with triggering code action at all.

So tbh I see them as different things 🤷🏿 I guess you could add similar functionality here but I'm not sure how lightbulb works under the hood. I guess it checks asynchronously if the `codeAction` handler returns any options on cursor hold and if so it will show a lightbulb in the UI of your choice e.g. signcolumn, virtual text etc

**Comment by glepnir** [comment]
I got it.implement it .

**Comment by glepnir** [comment]
And also can improve the diagnostic .when we jump to it if has code action available show sign and virtual text etc.

**Comment by cooper-anderson** [comment]
I know some plugins have a feature where when you jump to a diagnostic, underneath the description it shows `(quickfix available)` or something similar.

**Comment by glepnir** [comment]
I implemented it simply, but I found that the virtual text does not perform well, if the diagnostic virtualtext is turned on. The lightbulb icon will not be displayed on the  line,

**Comment by glepnir** [comment]
I think the best workflow  to fix the diagnosis is to jump to the wrong location and display the diagnosis information 
1. show code action available or not in diagnostic float window and show sign. 
2. show code action available or not use virtual text.

the first way is not conflict with open diagnostic virtual text.

**Comment by glepnir** [comment]
I think the virtual text of code action should only show in current line which has diagnostic  and keep only display one .when we move to new line. should clear the prev line virtual text and sign of code action. any idea?

**Comment by glepnir** [comment]
Yes, I tested the behavior in vscode. nvim-lighthub has some noise. When the cursor is on the wrong line, it displays the sign virtual text, but when the cursor moves to the correct line, the sign and virtual text at the previous position will not disappear. This is different from vscode. I am implementing a behavior similar to vscode. When leaving the wrong line. Then the sign and virtualtext at that position will disappear.

**Comment by glepnir** [comment]
Now works like this .

![finder](https://user-images.githubusercontent.com/41671631/110444358-e0b27000-80f7-11eb-90f7-44619ee68820.gif)

About float window. I think call this in diagnostic jump wil be useful. use diagnostic_goto_next. and it will show lighthub and flaotwindow to show diagnostic message.

about sign support it later . any idea?

**Comment by glepnir** [comment]
![finder](https://user-images.githubusercontent.com/41671631/110447985-bc589280-80fb-11eb-8474-eb71dd14aaa9.gif)

**Comment by akinsho** [comment]
@glepnir it looks like it stays in the signcolumn even when you navigate away from the line, which will get very noisy. Also regarding the floating window how about when there is no diagnostic just a code action will that show a separate window for code action lightbulb alone?

**Comment by akinsho** [comment]
Yep, that's quite common, the code action light bulb isn't just for errors it should show whenever there is an option to do something to the code. Here's a quick example using lightbulb in some code I'm working on.

![code_action_lsp](https://user-images.githubusercontent.com/22454918/110448909-8eeaf500-80b9-11eb-960b-9ca8a97c9185.gif)

**Comment by glepnir** [comment]
got this ..should be fix .now I ues the diagnostic to check code action .improve this later.

**Comment by glepnir** [comment]
@akinsho  can you test it now?

**Comment by akinsho** [comment]
Just tried in the same file as I did above with the most recent change and I never see a lightbulb on any of the same lines. I configured it as
```lua
    code_action_prompt = {
      enable = true,
      sign = false,
      virtual_text = true
    }

```

**Comment by akinsho** [comment]
It looks like all code actions stopped working as well so event pressing `:Lspsaga code_action` doesn't work where it used to before

**Comment by glepnir** [comment]
![Untitled](https://user-images.githubusercontent.com/41671631/110482231-67c90d80-8123-11eb-9468-be257d51c364.gif)

**Comment by lucax88x** [comment]
It's possible to disable this feature?

**Comment by glepnir** [comment]
set prompt enable. to false.

**Comment by akinsho** [comment]
@glepnir tried again and doesn't work in `lua` or `dart`. Can you try in some other filetype other than `go` and see if you can reproduce what I'm seeing.

![lsp_saga_codeaction](https://user-images.githubusercontent.com/22454918/110514447-34748780-80ff-11eb-885c-e6850af4abd3.gif)

From the GIF you can see that there *was* a code action available at the cursor position but no lightbulb appeared

**Comment by akinsho** [comment]
I've also noticed, the error below when I'm in vim files using `vimls`. Probably it doesn't support code action so trying to trigger the lightbulb breaks
```
E5108: Error executing lua .../pack/packer/opt/lspsaga.nvim/lua/lspsaga/codeaction.lua:60: bad argument #1 to 'next' (table expected, got number)
```

**Comment by glepnir** [comment]
can you test it now.

**Comment by nghialm269** [comment]
> I've also noticed, the error below when I'm in vim files using `vimls`. Probably it doesn't support code action so trying to trigger the lightbulb breaks
> 
> ```
> E5108: Error executing lua .../pack/packer/opt/lspsaga.nvim/lua/lspsaga/codeaction.lua:60: bad argument #1 to 'next' (table expected, got number)
> ```



> can you test it now.

@glepnir  I still can reproduce the issue with `vimls`.

**Comment by glepnir** [comment]
you need update plugin..

**Comment by nghialm269** [comment]
![image](https://user-images.githubusercontent.com/13317164/110571870-e0af8000-818a-11eb-9e14-14b23deaca85.png)
still can reproduce after updating to 38b2143b84a9bfec779bc8132a080fff53c9d120

**Comment by glepnir** [comment]
how about now?

**Comment by nghialm269** [comment]
> how about now?

It's fixed now, thank you :)

**Comment by glepnir** [comment]
I think now can work better .

**Comment by akinsho** [comment]
@glepnir I had a look and it works much better 👍🏿 and no longer seeing the error in vim files. One thing which I think but this is a matter of preference, the default seems to be to enable the signcolumn *and* the virtual text i think it should be one *or* the other I'm not sure it's useful to have the sign duplicated in two different areas by default 🤷🏿 just my opinion but think a lot of people will want to disable either one or the other

**Comment by tami5** [comment]
Yah two of them enabled in same time might not be a good default. I vote to use sign column as default since it's a sign.

**Comment by glepnir** [comment]
Use signcolumn  behave more like vscode :)
